### ISSUE TITLE
Tutorial for running atac.wdl with Caper

### ISSUE THREAD

**Issue by memathyer** [ENCODE-DCC/atac-seq-pipeline#146]
TL:DR: Is there a more thorough step-by-step tutorial for using caper to run the ATAC-seq pipeline?

I am attempting to install and use caper to run the atac-seq pipeline and I cannot get the test data to run using the provided instructions. Is there a more step-by-step tutorial for people who may not be familiar with bioinformatics pipelines?  I feel like there are a few steps skipped in the current instructions that may be obvious to people who do this regularly but I am not aware of.

Following the instructions I: 
installed caper
$ pip install caper

I then created a configuration file by generating a default file and used this to create my own configuration file to specify using local backend, specified a DB file (what is this? I couldn’t find a definition of this file), define an output directory and a singularity cache folder.
(configuration file attached)
[theirtestdata.config.pdf](https://github.com/ENCODE-DCC/atac-seq-pipeline/files/3503132/theirtestdata.config.pdf)

I then attempted to run the test data exactly as instructed

$ caper run atac.wdl -i examples/caper/ENCSR356KRQ_subsampled.json --deepcopy --use-singularity

This immediately errored because it couldn’t find the atac.wdl and the input files.  I know this may seem obvious, but a command explaining how to download the pipeline and associated data from the github, and where to put that data would be helpful, particularly for people non-bioinfomaticians attempting to use this pipeline.

So I downloaded the pipeline, and ran the above command in the folder with the pipeline

$ caper run atac.wdl -i examples/caper/ENCSR356KRQ_subsampled.json --deepcopy --use-singularity

(I did $ module load singularity first)

This gets to the build local singularity image step then dies

[Caper] building local singularity image:  docker://quay.io/encode-dcc/atac-seq-pipeline:v1.4.2
Traceback (most recent call last):
  File "/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/bin/caper", line 13, in <module>
    main()
  File "/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/lib/python3.7/site-packages/caper/caper.py", line 1125, in main
    c.run()
  File "/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/lib/python3.7/site-packages/caper/caper.py", line 204, in run
    input_file, tmp_dir)
  File "/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/lib/python3.7/site-packages/caper/caper.py", line 644, in __create_workflow_opts_json_file
    self.__build_singularity_image(singularity)
  File "/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/lib/python3.7/site-packages/caper/caper.py", line 941, in __build_singularity_image
    return check_call(cmd, env=env)
  File "/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/lib/python3.7/subprocess.py", line 342, in check_call
    retcode = call(*popenargs, **kwargs)
  File "/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/lib/python3.7/subprocess.py", line 323, in call
    with Popen(*popenargs, **kwargs) as p:
  File "/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/lib/python3.7/subprocess.py", line 775, in __init__
    restore_signals, start_new_session)
  File "/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/lib/python3.7/subprocess.py", line 1522, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'singularity': 'singularity'

I then tried to follow the instructions from the Caper README file
where I specify the configuration file 

$ caper -c ~/theirtestdata.conf run atac.wdl -i examples/local/ENCSR356KRQ_subsampled.json

The pipeline gets to the python $(which encode_trim_adapter.py) step before erroring out
Error file-

$ cat /scratch/mary.mathyer/caperruns/theirtestdata/atac/6bfa020f-e78e-4b80-86c1-77d4a77fafd4/call-trim_adapter/shard-0/attempt-2/execution/stderr.background 
which: no encode_trim_adapter.py in (/export/singularity-3.2.1/bin/:/act/jre1.8.0_40/bin:/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/bin:/export/Anaconda3-5.2.0/bin:/usr/local/bin:/usr/local/sbin:/usr/lib64/qt-3.3/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/sbin:/act/bin:/home/mary.mathyer/bin)
  File "/scratch/mary.mathyer/caperruns/theirtestdata/atac/6bfa020f-e78e-4b80-86c1-77d4a77fafd4/call-trim_adapter/shard-0/attempt-2/execution/write_tsv_09eb76710e46160a1fcfec451b776e44.tmp", line 1
    /scratch/mary.mathyer/caperruns/theirtestdata/atac/6bfa020f-e78e-4b80-86c1-77d4a77fafd4/call-trim_adapter/shard-0/attempt-2/inputs/543142024/ENCFF341MYG.subsampled.400.fastq.gz	/scratch/mary.mathyer/caperruns/theirtestdata/atac/6bfa020f-e78e-4b80-86c1-77d4a77fafd4/call-trim_adapter/shard-0/attempt-2/inputs/543142025/ENCFF248EJF.subsampled.400.fastq.gz
    ^
SyntaxError: invalid syntax
ln: accessing `/scratch/mary.mathyer/caperruns/theirtestdata/atac/6bfa020f-e78e-4b80-86c1-77d4a77fafd4/call-trim_adapter/shard-0/attempt-2/execution/R1/*.fastq.gz': No such file or directory
ln: accessing `/scratch/mary.mathyer/caperruns/theirtestdata/atac/6bfa020f-e78e-4b80-86c1-77d4a77fafd4/call-trim_adapter/shard-0/attempt-2/execution/R2/*.fastq.gz': No such file or directory

If I attempt to specify caper  to use a singularity container I have previously made and used successfully to run this test data without caper I get similar error messages to before-

$ caper -c ~/theirtestdata.conf run atac.wdl -i examples/local/ENCSR356KRQ_subsampled.json --singularity ~/.singularity/atac-seq-pipeline-v1.4.2.simg 
[Caper] building local singularity image:  /home/mary.mathyer/.singularity/atac-seq-pipeline-v1.4.2.simg
Traceback (most recent call last):
  File "/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/bin/caper", line 13, in <module>
    main()
  File "/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/lib/python3.7/site-packages/caper/caper.py", line 1125, in main
    c.run()
  File "/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/lib/python3.7/site-packages/caper/caper.py", line 204, in run
    input_file, tmp_dir)
  File "/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/lib/python3.7/site-packages/caper/caper.py", line 644, in __create_workflow_opts_json_file
    self.__build_singularity_image(singularity)
  File "/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/lib/python3.7/site-packages/caper/caper.py", line 941, in __build_singularity_image
    return check_call(cmd, env=env)
  File "/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/lib/python3.7/subprocess.py", line 342, in check_call
    retcode = call(*popenargs, **kwargs)
  File "/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/lib/python3.7/subprocess.py", line 323, in call
    with Popen(*popenargs, **kwargs) as p:
  File "/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/lib/python3.7/subprocess.py", line 775, in __init__
    restore_signals, start_new_session)
  File "/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/lib/python3.7/subprocess.py", line 1522, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'singularity': 'singularity'

Any suggestions on what I am doing wrong? Is there some sort of step I am missing or did I not install something correctly? Assistance would be greatly appreciated

**Comment by leepc12** [comment]
Thanks for your comments. We will try to improve documentation in the next release.

It looks like you don't have `singularity` on your system.

Check if the following works and if not then install singularity.
```
$ singularity exec docker://ubuntu:latest echo "hello world"
hello world
```

**Comment by memathyer** [comment]
I thought I had singularity installed since I have used it to run the pipeline without caper but that above command failed

$ singularity exec docker://ubuntu:latest echo "hello world"
INFO:    Converting OCI blobs to SIF format
INFO:    Starting build...
Getting image source signatures
Copying blob sha256:7413c47ba209e555018c4be91101d017737f24b0c9d1f65339b97a4da98acb2a
 25.45 MiB / 25.45 MiB [====================================================] 0s
Copying blob sha256:0fe7e7cbb2e88617d969efeeb3bd3125f7d309335c736a0525233ec2dc06aee1
 34.54 KiB / 34.54 KiB [====================================================] 0s
Copying blob sha256:1d425c98234572d4221a1ac173162c4279f9fdde4726ec22ad3c399f59bb7503
 847 B / 847 B [============================================================] 0s
Copying blob sha256:344da5c95cecd0f55238ce59b8469ee301056001ece2b769e9691b80f94f9f37
 162 B / 162 B [============================================================] 0s
Copying config sha256:5f40de30a14521a4e532d4a73021d2efe54e26309406841f266a4ebf52508528
 2.42 KiB / 2.42 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures
INFO:    Creating SIF file...
INFO:    Build complete: /home/mary.mathyer/.singularity/cache/oci-tmp/c303f19cfe9ee92badbbbd7567bc1ca47789f79303ddcef56f77687d4744cd7a/ubuntu_latest.sif
WARNING: underlay of /etc/localtime required more than 50 (65) bind mounts
FATAL: kernel too old


What version of Singularity does this pipeline need because I currently seem to have singularity-3.2.1 installed and loaded
$ which singularity
/export/singularity-3.2.1/bin/singularity

**Comment by leepc12** [comment]
The reason for this error is quite obvious. Pipeline couldn't find `singularity` executable in your `$PATH`. Is there some initization things defined in your `~/.bashrc` like `module load singularity`?

**Comment by memathyer** [comment]
hmm, I did module load singularity to add singularity to my path and the above test command still failed.

$ echo $PATH
/export/singularity-3.2.1/bin/:/act/jre1.8.0_40/bin:/home/mary.mathyer/.conda/envs/Anaconda3-5.2.0/bin:/export/Anaconda3-5.2.0/bin:/usr/local/bin:/usr/local/sbin:/usr/lib64/qt-3.3/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/sbin:/act/bin:/home/mary.mathyer/bin
(Anaconda3-5.2.0) [mary.mathyer@node06 ~]$ singularity exec docker://ubuntu:latest echo "hello world"
WARNING: underlay of /etc/localtime required more than 50 (65) bind mounts
FATAL: kernel too old

running the test data using 
$ caper run atac.wdl -i examples/caper/ENCSR356KRQ_subsampled.json --deepcopy --use-singularity
Still produces the FileNotFoundError: [Errno 2] No such file or directory: 'singularity': 'singularity'

This version of singularity works when I run the atac.wdl pipeline without caper.

**Comment by memathyer** [comment]
I have spoken with individuals more familiar with my system it looks like we are running a relatively old kernel. The test Docker image you're having me test with is incompatible with our underlying system's kernel.  Could this also be why I have been unsuccessful with running caper?  If so, do you have any workarounds that may be compatible with an older kernel? Thanks

**Comment by leepc12** [comment]
Pipeline just use `singularity exec` to build a container and run command lines. I think I have less knowledge than you and your admins about singularity kernel things. Instead, can you try with Conda (4.6.x)?

**Comment by leepc12** [comment]
Closing this due to long inactivity
